python/demo/demo_pml/demo_pml.py

```diff
@@ -0,0 +1,606 @@
+
+model = MPI.COMM_WORLD.bcast(model, root=0)
+domain, cell_tags, facet_tags = gmshio.model_to_mesh(
+    model, MPI.COMM_WORLD, 0, gdim=2)
+
+gmsh.finalize()
+MPI.COMM_WORLD.barrier()
+# -
+
+# We visualize the mesh and subdomains with [PyVista](https://docs.pyvista.org/)
+
+if have_pyvista:
+    topology, cell_types, geometry = plot.create_vtk_mesh(domain, 2)
+    grid = pyvista.UnstructuredGrid(topology, cell_types, geometry)
+    pyvista.set_jupyter_backend("pythreejs")
+    plotter = pyvista.Plotter()
+    num_local_cells = domain.topology.index_map(domain.topology.dim).size_local
+    grid.cell_data["Marker"] = \
+        cell_tags.values[cell_tags.indices < num_local_cells]
+    grid.set_active_scalars("Marker")
+    plotter.add_mesh(grid, show_edges=True)
+    plotter.view_xy()
+    if not pyvista.OFF_SCREEN:
+        plotter.show(interactive=True)
+    else:
+        pyvista.start_xvfb()
+        figure = plotter.screenshot("wire_mesh_pml.png",
+                                    window_size=[800, 800])
+
+# We observe five different subdomains: one for the gold
+# wire (`au_tag`), one for the background medium (`bkg_tag`), one for the
+# PML corners (`pml_tag`), one for the PML rectangles along $x$
+# (`pml_tag + 1`), and one for the PML rectangles along $y$ (`pml_tag + 2`).
+# These different PML regions have different coordinate transformation,
+# as specified here below:
+#
+# \begin{align}
+# \text{PML}_\text{corners} \rightarrow \mathbf{r}^\prime & = (x^\prime, y^\prime) \\
+# \text{PML}_\text{rectangles along x} \rightarrow
+#                                       \mathbf{r}^\prime & = (x^\prime, y) \\
+# \text{PML}_\text{rectangles along y} \rightarrow
+#                                       \mathbf{r}^\prime & = (x, y^\prime).
+# \end{align}
+#
+# Now we define some other problem specific parameters:
+
+wl0 = 0.4  # Wavelength of the background field
+n_bkg = 1  # Background refractive index
+eps_bkg = n_bkg**2  # Background relative permittivity
+k0 = 2 * np.pi / wl0  # Wavevector of the background field
+theta = 0  # Angle of incidence of the background field
+
+# We use a degree 3
+# [Nedelec (first kind)](https://defelement.com/elements/nedelec1.html)
+# element to represent the electric field:
+#
+
+degree = 3
+curl_el = ufl.FiniteElement("N1curl", domain.ufl_cell(), degree)
+V = fem.FunctionSpace(domain, curl_el)
+
+# Next, we interpolate $\mathbf{E}_b$ into the function space $V$, define our
+# trial and test function, and the integration domains:
+
+# +
+Eb = fem.Function(V)
+f = partial(background_field, theta, n_bkg, k0)
+Eb.interpolate(f)
+
+# Definition of Trial and Test functions
+Es = ufl.TrialFunction(V)
+v = ufl.TestFunction(V)
+
+# Definition of 3d fields
+Es_3d = ufl.as_vector((Es[0], Es[1], 0))
+v_3d = ufl.as_vector((v[0], v[1], 0))
+
+# Measures for subdomains
+dx = ufl.Measure("dx", domain, subdomain_data=cell_tags)
+dDom = dx((au_tag, bkg_tag))
+dPml_xy = dx(pml_tag)
```

Comment:
Here you're defining three domains for the three parts of the PML, which you then use in your definition of the weak form. Was there a particular reason why you have done it like that instead of defining cellwise-constant `fem.Function`s over the entire domain? I'm asking because in my own implementation of an UPML, I'm using `fem.Function`s.

For `eps` you are even defining a `fem.Function`, but you are not using it in the PML region(s).

Reply:
The main reason is that we wanted to explicitly show that we have three different transformations in the six different regions of the PML.

Besides, I think that using `fem.Function` would make the computation a bit slower, since you need to interpolate the function over a certain function space (but I might be wrong).

Reply:
> Besides, I think that using fem.Function would make the computation a bit slower, since you need to interpolate the function over a certain function space (but I might be wrong).

I agree with you that it would probably be slower, but I haven't verified if using `fem.Function` has a measurable impact. In my application the user has to provide a permittivity function and a permeability function for the entire domain (including the PML). The solver in my application is not aware that there is a PML in the geometry (In 1D the outgoing boundary condition can be imposed at the boundary of your geometry, i.e., a PML is not needed.). I think both approaches (yours and mine) have their pros and cons.